File: Binary Search/875. Koko Eating Bananas/875. Koko Eating Bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function(piles, h) {
    let low = 1
    let high = Math.max(...piles)

    while (low <= high) {
        const k = Math.floor((low + high) / 2)
        const curr_time = piles.reduce((a, c) => a + Math.ceil(c / k), 0)
        if (curr_time <= h) high = k - 1
        else low = k + 1
    }

    return low
};

// Time: O(n log max(piles))
// Space: O(1)

File: Binary Search/875. Koko Eating Bananas/875. Koko Eating Bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low = 1
        high = max(piles)

        while (low <= high):
            k = (low + high) // 2
            curr_time = 0
            for b in piles:
                curr_time += ceil(b / k)

            if curr_time <= h:
                high = k - 1
            else:
                low = k + 1
        
        return low

# Time: O(n log max(piles))
# Space: O(1)